crates/sui-framework/packages/bridge/sources/treasury.move

```diff
@@ -2,86 +2,217 @@
+    use sui::math;
+    use sui::object;
+    use sui::object_bag::{Self, ObjectBag};
+    use sui::package;
+    use sui::package::UpgradeCap;
+    use sui::vec_map;
+    use sui::vec_map::VecMap;
 
     const EUnsupportedTokenType: u64 = 0;
-    const ENotSystemAddress: u64 = 1;
+    const EInvalidUpgradeCap: u64 = 1;
+    const ETokenSupplyNonZero: u64 = 2;
+
+    #[test_only]
+    const USD_VALUE_MULTIPLIER: u64 = 10000; // 4 DP accuracy
 
     public struct BridgeTreasury has store {
-        treasuries: ObjectBag
+        // token treasuries, values are TreasuryCaps for native bridge V1, it can also store Vaults for native tokens in future release.
+        treasuries: ObjectBag,
+        supported_tokens: VecMap<TypeName, BridgeTokenMetadata>,
+        // Mapping token id to type name
+        id_token_type_map: VecMap<u8, TypeName>,
+        // Bag for storing potential new token waiting to be approved
+        waiting_room: Bag
     }
 
-    public fun token_id<T>(): u8 {
-        let coin_type = type_name::get<T>();
-        if (coin_type == type_name::get<BTC>()) {
-            1
-        } else if (coin_type == type_name::get<ETH>()) {
-            2
-        } else if (coin_type == type_name::get<USDC>()) {
-            3
-        } else if (coin_type == type_name::get<USDT>()) {
-            4
-        } else {
-            abort EUnsupportedTokenType
-        }
+    public struct BridgeTokenMetadata has store, copy, drop {
+        id: u8,
+        decimal_multiplier: u64,
+        notional_value: u64,
+        native_token: bool
     }
 
-    public fun decimal_multiplier<T>(): u64 {
-        let coin_type = type_name::get<T>();
-        if (coin_type == type_name::get<BTC>()) {
-            btc::multiplier()
-        } else if (coin_type == type_name::get<ETH>()) {
-            eth::multiplier()
-        } else if (coin_type == type_name::get<USDC>()) {
-            usdc::multiplier()
-        } else if (coin_type == type_name::get<USDT>()) {
-            usdt::multiplier()
+    public struct ForeignTokenRegistration has store {
+        type_name: TypeName,
+        uc: UpgradeCap,
+        decimal: u8,
+    }
+
+    public struct UpdateTokenPriceEvent has copy, drop {
+        token_id: u8,
+        new_price: u64,
+    }
+
+    public struct NewTokenEvent has copy, drop {
+        token_id: u8,
+        type_name: TypeName,
+        native_token: bool
+    }
+
+    public struct TokenRegistrationEvent has copy, drop {
+        type_name: TypeName,
+        decimal:u8,
+        native_token: bool
+    }
+
+    public(package) fun register_foreign_token<T>(self: &mut BridgeTreasury, tc: TreasuryCap<T>, uc: UpgradeCap, metadata: &CoinMetadata<T>) {
+        // Make sure TreasuryCap has not been minted before.
+        assert!(coin::total_supply(&tc) == 0, ETokenSupplyNonZero);
+        let type_name = type_name::get<T>();
+        let coin_address = address::from_ascii_bytes(ascii::as_bytes(&type_name::get_address(&type_name)));
```

Comment:
this is the problematic line that is failing the tests.
I have to say that I am truly baffled as to what is going on, the logic in our framework verification, and how to fix this (which at the moment I do not know).
First off, as amusement, comment the call to `from_ascii_bytes` (e.g. `let coin_address = @0xABCD; // address::from_ascii_bytes(ascii::as_bytes(&type_name::get_address(&type_name)));`) and the tests will pass.
I wish I trusted the print line I put there which reported the missing function being `address::from_ascii_bytes`, but I did not believe it.
Of course the function is there but it was introduced in this PR https://github.com/MystenLabs/sui/pull/15427 which was committed on Jan 12 and likely made available after protocol config 35 which is the version before that PR. 
The tests in question run at protocol version 31 (`test_create_randomness_state_object`) and 34 (`test_coin_deny_list_creation`) and when verifying the bridge they choke with that error.
What I think is happening (and I have not verified it yet, I'll do it tomorrow) is that another module loads the `address` module at the given version (which does not contain the function) and when verification for the bridge happens (specifically the `treasury` module), it will not find that function because the loaded module does not have that.

I am not clear why we are verifying the latest framework even though we are not loading it. And it makes for a real mess in loading and dependency hell.
Maybe that operation should be done in its own VM, I do not know right now.

I don't know what to say. If you can find a way not to use that function that is an option (though not sure it is possible). I'd also be fine committing as is and we'll fix it next week, if others are ok.
Otherwise we'll see but at this point I have doubts I can fix it over the weekend.

Also it is something to run by @amnn and @mystenmark at the very least.